File: cpp/src/Options.cpp
#include <Options.h>
#include <Utils.h>

#include <cstdio>

Options::Options() :
    m_isNumPartitionsSet(false),
    m_isASet(false),
    m_isBSet(false),
    m_isFxSet(false)
{
}

bool Options::parse(int argc, char **argv) {
    // convert every argument to std::string for easy of equality testing
    std::vector<std::string> args;
    for (int i = 0; i < argc; i++) {
        args.push_back(std::string(argv[i]));
    }

    bool valueToReturn = true;

    for (size_t i = 1; i < args.size(); i += 2) {
        // take an argument and the next and asume is the command and it
        // respective argument
        std::string command = args[i];
        if (i+1 >= args.size()) {
            continue;
        }
        std::string arg = args[i+1];

        bool ok = false;

        if (command == "--num_partitions") {
            m_numPartitions = Utils::string2int(arg, &ok);
            if (ok) {
                m_isNumPartitionsSet = true;
            } else {
                std::cout << "Can't parse the argument for the "
                          << "--num_partitions command line option: "
                          << arg << std::endl;
                valueToReturn = false;
            }
        } else if (command == "--a") {
            m_a = Utils::string2double(arg);
            m_isASet = true;
        } else if (command == "--b") {
            m_b = Utils::string2double(arg);
            m_isBSet = true;
        } else if (command == "--function") {
            m_fx = Utils::string2int(arg);
            if (m_fx < 0 || m_fx > 9) {
                std::cout << "The argument for the --function command line "
                          << "option is out of range [0,9]: " << arg
                          << std::endl;
                valueToReturn = false;
            } else {
                m_isFxSet = true;
            }
        } else {
            printf("Command unknown!: %s\n", command.c_str());
            fflush(stdout);
            valueToReturn = false;
        }
    }

    if (!m_isNumPartitionsSet) {
        printf("Make sure to add the --num_partitions option! Error\n");
        fflush(stdout);
        valueToReturn = false;
    }

    if (!m_isASet) {
        printf("Make sure to add the --a option! Error\n");
        fflush(stdout);
        valueToReturn = false;
    }

    if (!m_isBSet) {
        printf("Make sure to add the --b option! Error\n");
        fflush(stdout);
        valueToReturn = false;
    }

    if (!m_isFxSet) {
        printf("Make sure to add the --function option ([0-9])! Error\n");
        fflush(stdout);
        valueToReturn = false;
    }

    return valueToReturn;
}

File: csharp/Options.cs
using System;

public class Options {
	public bool isNumPartitionsSet;
    public bool isASet;
    public bool isBSet;
    public bool isFxSet;

    public int numPartitions;
    public double a;
    public double b;
    public int fx;

    public bool parse(String []argv) {
        bool valueToReturn = true;

        for (int i = 0; i < argv.Length; i++) {
            String command = argv[i];
            if (i+1 >= argv.Length) {
                continue;
            }
            String arg = argv[i+1];

            if ("--num_partitions".Equals(command)) {
                try {
                    numPartitions = Convert.ToInt32(arg);
                    isNumPartitionsSet = true;
                } catch (FormatException e) {
                    Console.WriteLine("Can't parse the argument for --num_partitions");
                    Console.WriteLine(e.Message);
                    valueToReturn = false;
                }
            } else if ("--a".Equals(command)) {
                try {
                    a = Convert.ToDouble(arg);
                    isASet = true;
                } catch (FormatException e) {
                    Console.WriteLine("Can't parse the argument for --a");
                    Console.WriteLine(e.Message);
                    valueToReturn = false;
                }
            } else if ("--b".Equals(command)) {
                try {
                    b = Convert.ToDouble(arg);
                    isBSet = true;
                } catch (FormatException e) {
                    Console.WriteLine("Can't parse the argument for --b");
                    Console.WriteLine(e.Message);
                    valueToReturn = false;
                }
            } else if ("--function".Equals(command)) {
                try {
                    fx = Convert.ToInt32(arg);
                    isFxSet = true;
                } catch (FormatException e) {
                    Console.WriteLine("Can't parse the argument for --function");
                    Console.WriteLine(e.Message);
                    valueToReturn = false;
                }
            }
        }

        if (!isNumPartitionsSet) {
            Console.WriteLine("Make sure to add the --num_partitions option! Error");
            valueToReturn = false;
        }

        if (!isASet) {
            Console.WriteLine("Make sure to add the --a option! Error");
            valueToReturn = false;
        }

        if (!isBSet) {
            Console.WriteLine("Make sure to add the --b option! Error");
            valueToReturn = false;
        }

        if (!isFxSet) {
 	        Console.WriteLine("Make sure to add the --function option ([0-9])! Error");
            valueToReturn = false;
        }

        return valueToReturn;
    }
}